File: app/src/main/java/com/quigglesproductions/secureimageviewer/dagger/hilt/module/RequestModule.java
package com.quigglesproductions.secureimageviewer.dagger.hilt.module;

import com.google.gson.Gson;
import com.quigglesproductions.secureimageviewer.appauth.RequestServiceNotConfiguredException;
import com.quigglesproductions.secureimageviewer.apprequest.RequestManager;
import com.quigglesproductions.secureimageviewer.dagger.hilt.annotations.RequestServiceClient;
import com.quigglesproductions.secureimageviewer.retrofit.RequestService;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.android.components.ActivityComponent;
import dagger.hilt.components.SingletonComponent;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

@Module
@InstallIn(ActivityComponent.class)
public class RequestModule {

    @Provides
    public static RequestService provideReqService(Gson gson, @RequestServiceClient OkHttpClient client){
        Retrofit retrofit = null;
        //try {
            retrofit = new Retrofit.Builder()
                    //.baseUrl(RequestManager.getInstance().getUrlManager().getBaseUrlString())
                    .baseUrl("https://quigleyserver.ddns.net:14500/api/")
                    .addConverterFactory(GsonConverterFactory.create(gson))
                    .client(client)
                    .build();
        //} catch (RequestServiceNotConfiguredException e) {
        //    throw new RuntimeException(e);
        //}

        return retrofit.create(RequestService.class);

    }
}

File: app/src/main/java/com/quigglesproductions/secureimageviewer/dagger/hilt/module/RequestModule.kt
package com.quigglesproductions.secureimageviewer.dagger.hilt.module

import com.google.gson.Gson
import com.quigglesproductions.secureimageviewer.dagger.hilt.annotations.AuthServiceClient
import com.quigglesproductions.secureimageviewer.dagger.hilt.annotations.DownloadServiceClient
import com.quigglesproductions.secureimageviewer.dagger.hilt.annotations.RequestServiceClient
import com.quigglesproductions.secureimageviewer.retrofit.DeviceRegistrationRequestService
import com.quigglesproductions.secureimageviewer.retrofit.DownloadService
import com.quigglesproductions.secureimageviewer.retrofit.IdRequestService
import com.quigglesproductions.secureimageviewer.retrofit.ModularRequestService
import com.skydoves.retrofit.adapters.paging.PagingCallAdapterFactory
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

@Module
@InstallIn(SingletonComponent::class)
object RequestModule {

    @Provides
    fun provideModularReqService(
        gson: Gson?,
        @RequestServiceClient client: OkHttpClient
    ): ModularRequestService {
        var retrofit: Retrofit? = null
        //try {
        retrofit =
            Retrofit.Builder() //.baseUrl(RequestManager.getInstance().getUrlManager().getBaseUrlString())
                .baseUrl("https://quigleyserver.ddns.net:14500/api/")
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(PagingCallAdapterFactory.create())
                .client(client)
                .build()
        //} catch (RequestServiceNotConfiguredException e) {
        //    throw new RuntimeException(e);
        //}
        return retrofit.create(ModularRequestService::class.java)
    }

    @Provides
    fun provideDownloadService(
        gson: Gson?,
        @DownloadServiceClient client: OkHttpClient
    ): DownloadService {
        var retrofit: Retrofit? = null
        //try {
        retrofit =
            Retrofit.Builder() //.baseUrl(RequestManager.getInstance().getUrlManager().getBaseUrlString())
                .baseUrl("https://quigleyserver.ddns.net:14500/api/")
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(PagingCallAdapterFactory.create())
                .client(client)
                .build()
        //} catch (RequestServiceNotConfiguredException e) {
        //    throw new RuntimeException(e);
        //}
        return retrofit.create(DownloadService::class.java)
    }

    @Provides
    fun provideRegistrationReqService(
        gson: Gson?,
        @AuthServiceClient client: OkHttpClient?
    ): DeviceRegistrationRequestService {
        var retrofit: Retrofit? = null
        //try {
        retrofit =
            Retrofit.Builder() //.baseUrl(RequestManager.getInstance().getUrlManager().getBaseUrlString())
                .baseUrl("https://quigleyserver.ddns.net:14500/api/")
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(PagingCallAdapterFactory.create())
                .client(client)
                .build()
        //} catch (RequestServiceNotConfiguredException e) {
        //    throw new RuntimeException(e);
        //}
        return retrofit.create(DeviceRegistrationRequestService::class.java)
    }

    @Provides
    fun provideIdReqService(
        gson: Gson?,
        @RequestServiceClient client: OkHttpClient?
    ): IdRequestService {
        var retrofit: Retrofit? = null
        //try {
        retrofit =
            Retrofit.Builder() //.baseUrl(RequestManager.getInstance().getUrlManager().getBaseUrlString())
                .baseUrl("https://quigleyid.ddns.net/v2/oauth")
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(PagingCallAdapterFactory.create())
                .client(client)
                .build()
        //} catch (RequestServiceNotConfiguredException e) {
        //    throw new RuntimeException(e);
        //}
        return retrofit.create(IdRequestService::class.java)
    }
}
